Firestore/core/src/firebase/firestore/core/event_manager.mm

```diff
@@ -67,30 +75,60 @@
   }
 }
 
+void EventManager::AddSnapshotsInSyncListener(
+    const std::shared_ptr<EventListener<Empty>>& listener) {
+  snapshots_in_sync_listeners_.insert(listener);
+  listener->OnEvent(Empty());
+}
+
+void EventManager::RemoveSnapshotsInSyncListener(
+    const std::shared_ptr<EventListener<Empty>>& listener) {
+  snapshots_in_sync_listeners_.erase(listener);
+}
+
 void EventManager::HandleOnlineStateChange(model::OnlineState online_state) {
+  bool raised_event = false;
   online_state_ = online_state;
 
   for (auto&& kv : queries_) {
     QueryListenersInfo& info = kv.second;
     for (auto&& listener : info.listeners) {
-      listener->OnOnlineStateChanged(online_state_);
+      if (listener->OnOnlineStateChanged(online_state_)) {
+        raised_event = true;
+      }
     }
   }
+  if (raised_event) {
+    RaiseSnapshotsInSyncEvent();
+  }
+}
+
+void EventManager::RaiseSnapshotsInSyncEvent() {
+  for (const std::shared_ptr<EventListener<Empty>>& listener :
```

Comment:
nit: since this is such a huge type, consider making this `const auto&` instead.